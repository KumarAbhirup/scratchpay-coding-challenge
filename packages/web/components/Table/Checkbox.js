import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { Checkbox as SemanticCheckbox } from 'semantic-ui-react'

/**
 * @name Checkbox
 *
 * @prop {Boolean} isChecked
 * @prop {Function} onChange
 */
export default function Checkbox(props) {
  const { isChecked, onChange } = props
  const [status, setStatus] = useState(isChecked)

  return (
    <SemanticCheckbox
      toggle
      checked={status}
      onChange={() => {
        setStatus(prevStatus => !prevStatus)

        /**
         * States in React are one-step-behind when passed to Parent Components.
         * So, we use `!status` instead of `status`
         *
         * @see https://stackoverflow.com/questions/28773839/react-form-onchange-setstate-one-step-behind
         *
         * `onChange` function reports to Parent Component.
         */
        if (onChange) onChange(!status)
      }}
    />
  )
}

Checkbox.propTypes = {
  isChecked: PropTypes.bool,
  onChange: PropTypes.func,
}
