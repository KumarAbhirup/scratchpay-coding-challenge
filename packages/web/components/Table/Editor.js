import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'

import { isEmailProper, isDuplicate } from '../../lib/utils'

/**
 * @name Editor
 *
 * @prop {String} name
 * @prop {String} value
 * @prop {String} placeholder
 * @prop {Function} onChange
 */
export default function Editor(props) {
  const { name, value, placeholder, onChange, emailsOfAllUsers } = props

  const [shouldShowTextarea, setShowTextarea] = useState(false) // State that decides whether the input field to edit should be shown or not.
  const [editorValue, setEditorValue] = useState(value) // Value of the controlled input component.

  // If the value is empty, then show the textarea. Else, allow the toggle.
  const decideWhetherToggleTextarea = () => {
    if (editorValue.length !== 0) {
      setShowTextarea(prevState => !prevState)
    } else setShowTextarea(true)
  }

  // If the input is empty, show the textarea instead of text.
  useEffect(() => {
    // Self invoking function.
    ;(() => {
      if (editorValue.length === 0) {
        setShowTextarea(true)
      }
    })()
  }, [editorValue])

  return (
    <div
      className="wrap-cell"
      onDoubleClickCapture={decideWhetherToggleTextarea}
    >
      {shouldShowTextarea ? (
        <input
          type={name === 'email' ? 'email' : 'text'}
          name={name}
          value={editorValue}
          placeholder={placeholder}
          onChange={e => {
            const valueToBe = e.target.value
            setEditorValue(valueToBe)

            // Report the change to parent component.
            if (onChange) onChange(e, valueToBe)
          }}
          onKeyDownCapture={e => {
            // Save the input when Enter or Tab key is pressed.
            if (e.key === 'Enter' || e.key === 'Tab') {
              decideWhetherToggleTextarea()
            }
          }}
        />
      ) : (
        value
      )}

      {/* If it is an email field, and if it is not empty, and if email isn't proper, then show the error. */}
      {name === 'email' &&
        editorValue.length !== 0 &&
        !isEmailProper(editorValue) && (
          <p className="error">Please write a valid email.</p>
        )}

      {/* If it is an email field, and if it is not empty, and if email is duplicate, then show the error. */}
      {name === 'email' &&
        editorValue.length !== 0 &&
        isDuplicate(emailsOfAllUsers, editorValue) && (
          <p className="error">Duplicate email.</p>
        )}
    </div>
  )
}

Editor.propTypes = {
  name: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  onChange: PropTypes.func,
  emailsOfAllUsers: PropTypes.array.isRequired,
}
