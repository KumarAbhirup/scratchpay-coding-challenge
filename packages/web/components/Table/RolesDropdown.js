import React from 'react'
import PropTypes from 'prop-types'
import { Dropdown } from 'semantic-ui-react'

export const roleOptions = [
  {
    key: 'doctor',
    text: 'Doctor',
    value: 'doctor',
    image: { avatar: true, src: '/images/doctor.png' },
  },
  {
    key: 'admin',
    text: 'Admin',
    value: 'admin',
    image: { avatar: true, src: '/images/admin.png' },
  },
  {
    key: 'accountant',
    text: 'Accoutant',
    value: 'accountant',
    image: { avatar: true, src: '/images/accountant.png' },
  },
]

export default function RolesDropdown(props) {
  const { activeRole } = props

  return (
    <Dropdown
      placeholder="Select Role"
      fluid
      selection
      options={roleOptions}
      value={activeRole}
    />
  )
}

RolesDropdown.propTypes = {
  activeRole: PropTypes.oneOf(roleOptions.map(role => role.value)),
}
