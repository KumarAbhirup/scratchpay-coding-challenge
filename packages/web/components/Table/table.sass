table.ui.celled
  border-top: 3px solid #245CA6
  font-family: 'Muli', "SF Pro Text", "SF Pro Icons", "Helvetica Neue", "Helvetica", "Arial", sans-serif

  button
    font-family: 'Muli', "SF Pro Text", "SF Pro Icons", "Helvetica Neue", "Helvetica", "Arial", sans-serif

  div.wrap-cell
    padding: 5px
    // border: 2px solid red
    overflow-y: auto

    textarea
      resize: none
      width: 100%

  p.table-tip
    color: #838383