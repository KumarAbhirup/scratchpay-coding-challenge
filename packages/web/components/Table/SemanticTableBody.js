import React from 'react'
import PropTypes from 'prop-types'
import { Table as SemanticTable } from 'semantic-ui-react'

import RolesDropdown, { roleOptions } from './RolesDropdown'
import Checkbox from './Checkbox'
import Editor from './Editor'

/**
 * @name SemanticTableBody
 *
 * @prop {Array<Object>} users
 * @prop {Function} onDataChange
 */
export default function SemanticTableBody(props) {
  const { users, onDataChange } = props

  // Called when text in the input field is changed.
  const handleEditorChange = (e, value, userIndex) => {
    const data = users // Modify `users` without mutating the original variable.
    data[userIndex][e.target.name] = value

    // Everytime there is a change, report to the parent component.
    onDataChange([...data])
  }

  const emails = users.map(user => user.email)

  return (
    <SemanticTable.Body>
      {users && // To avoid 'Cannot read property 'map' of undefined'.
        users.map(
          (user, userIndex) =>
            user && (
              <SemanticTable.Row key={user.id}>
                <SemanticTable.Cell>
                  <button
                    type="button"
                    onClick={() => {
                      const data = users

                      // Delete the user from the array
                      data.splice(userIndex, 1)

                      // Report to the parent component
                      onDataChange([...data])
                    }}
                  >
                    ‚ùå
                  </button>
                </SemanticTable.Cell>
                <SemanticTable.Cell>{user.id}</SemanticTable.Cell>
                <SemanticTable.Cell>
                  <Checkbox
                    isChecked={user.active}
                    onChange={isChecked => {
                      const data = users

                      // Report the change to the parent component everytime the Checkbox is toggled.
                      data[userIndex].active = isChecked
                      onDataChange([...data])
                    }}
                  />
                </SemanticTable.Cell>
                <SemanticTable.Cell>
                  <Editor
                    name="firstName"
                    value={user.firstName}
                    placeholder="First Name"
                    onChange={(e, value) =>
                      handleEditorChange(e, value, userIndex)
                    }
                    emailsOfAllUsers={emails}
                  />
                </SemanticTable.Cell>
                <SemanticTable.Cell>
                  <Editor
                    name="lastName"
                    value={user.lastName}
                    placeholder="Last Name"
                    onChange={(e, value) =>
                      handleEditorChange(e, value, userIndex)
                    }
                    emailsOfAllUsers={emails}
                  />
                </SemanticTable.Cell>
                <SemanticTable.Cell>
                  <Editor
                    name="email"
                    value={user.email}
                    placeholder="E-mail"
                    onChange={(e, value) =>
                      handleEditorChange(e, value, userIndex)
                    }
                    emailsOfAllUsers={emails}
                  />
                </SemanticTable.Cell>
                <SemanticTable.Cell>
                  <RolesDropdown
                    activeRole={user.role}
                    onChange={role => {
                      const data = users

                      // Report the change to the parent component everytime the Role is changed.
                      data[userIndex].role = role
                      onDataChange([...data])
                    }}
                  />
                </SemanticTable.Cell>
              </SemanticTable.Row>
            )
        )}
    </SemanticTable.Body>
  )
}

SemanticTableBody.propTypes = {
  users: PropTypes.arrayOf(
    PropTypes.shape(
      {
        id: PropTypes.number.isRequired,
        active: PropTypes.bool.isRequired,
        firstName: PropTypes.string.isRequired,
        lastName: PropTypes.string.isRequired,
        email: PropTypes.string.isRequired,
        role: PropTypes.oneOf(roleOptions.map(role => role.value)),
      }.isRequired
    )
  ),
  onDataChange: PropTypes.func,
}
