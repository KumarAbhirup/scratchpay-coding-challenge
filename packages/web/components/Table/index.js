import React, { useState } from 'react'
import { Button, Icon, Table as SemanticTable } from 'semantic-ui-react'

import './table.sass'
import SemanticTableBody from './SemanticTableBody'

const defaultUsers = [
  {
    id: 1,
    active: true,
    firstName: 'Kumar',
    lastName: 'Abhirup',
    email: 'kumarabhirup5@gmail.com',
    role: 'admin',
  },
  {
    id: 2,
    active: false,
    firstName: 'Yogita',
    lastName: 'Patil',
    email: 'yogita@iqubex.com',
    role: 'accountant',
  },
]

export default function Table() {
  const [users, setUsers] = useState(defaultUsers)

  return (
    <SemanticTable compact definition color="blue">
      <SemanticTable.Header fullWidth>
        <SemanticTable.Row>
          <SemanticTable.HeaderCell />
          <SemanticTable.HeaderCell>ID</SemanticTable.HeaderCell>
          <SemanticTable.HeaderCell>Status</SemanticTable.HeaderCell>
          <SemanticTable.HeaderCell>First Name</SemanticTable.HeaderCell>
          <SemanticTable.HeaderCell>Last Name</SemanticTable.HeaderCell>
          <SemanticTable.HeaderCell>E-mail address</SemanticTable.HeaderCell>
          <SemanticTable.HeaderCell>Role</SemanticTable.HeaderCell>
        </SemanticTable.Row>
      </SemanticTable.Header>

      <SemanticTableBody users={users} onDataChange={data => setUsers(data)} />

      <SemanticTable.Footer fullWidth>
        <SemanticTable.Row>
          <SemanticTable.HeaderCell />
          <SemanticTable.HeaderCell colSpan="6">
            <Button
              floated="right"
              icon
              labelPosition="left"
              primary
              size="small"
              onClick={() => {
                const data = [
                  ...users,
                  {
                    id: 3, // some random temporary number
                    active: false,
                    firstName: '',
                    lastName: '',
                    email: '',
                    role: 'admin',
                  },
                ]

                setUsers([...data])
              }}
            >
              <Icon name="user" /> Add User
            </Button>
            <p className="table-tip">
              Double-click a cell to edit, Hit enter to save.
            </p>
          </SemanticTable.HeaderCell>
        </SemanticTable.Row>
      </SemanticTable.Footer>
    </SemanticTable>
  )
}
