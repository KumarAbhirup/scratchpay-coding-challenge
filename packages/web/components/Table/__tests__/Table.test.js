import React from 'react'
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json'

import Table from '..'

describe('<Table />', () => {
  it('renders', () => {
    shallow(<Table />)
  })

  it('matches snapshot', () => {
    const wrapper = shallow(<Table />)

    expect(toJSON(wrapper)).toMatchSnapshot()
  })
})
