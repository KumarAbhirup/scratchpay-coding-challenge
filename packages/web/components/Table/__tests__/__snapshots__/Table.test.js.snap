// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> matches snapshot 1`] = `
<Table
  as="table"
  color="blue"
  compact={true}
  definition={true}
>
  <TableHeader
    as="thead"
    fullWidth={true}
  >
    <TableRow
      as="tr"
      cellAs="td"
    >
      <TableHeaderCell
        as="th"
      />
      <TableHeaderCell
        as="th"
      >
        ID
      </TableHeaderCell>
      <TableHeaderCell
        as="th"
      >
        Status
      </TableHeaderCell>
      <TableHeaderCell
        as="th"
      >
        First Name
      </TableHeaderCell>
      <TableHeaderCell
        as="th"
      >
        Last Name
      </TableHeaderCell>
      <TableHeaderCell
        as="th"
      >
        E-mail address
      </TableHeaderCell>
      <TableHeaderCell
        as="th"
      >
        Role
      </TableHeaderCell>
    </TableRow>
  </TableHeader>
  <SemanticTableBody
    onDataChange={[Function]}
    users={Array []}
  />
  <TableFooter
    as="tfoot"
    fullWidth={true}
  >
    <TableRow
      as="tr"
      cellAs="td"
    >
      <TableHeaderCell
        as="th"
      />
      <TableHeaderCell
        as="th"
        colSpan="6"
      >
        <Button
          as="button"
          floated="right"
          icon={true}
          labelPosition="left"
          onClick={[Function]}
          primary={true}
          size="small"
        >
          <Icon
            as="i"
            name="user"
          />
           Add User
        </Button>
        <p
          className="table-tip"
        >
          Double-click a cell to toggle the editor. Hit enter to save.
        </p>
      </TableHeaderCell>
    </TableRow>
  </TableFooter>
</Table>
`;
